Corona-Warn-App/src/main/java/de/rki/coronawarnapp/covidcertificate/signature/core/DscSignatureValidator.kt

```diff
@@ -1,19 +1,184 @@
+import de.rki.coronawarnapp.covidcertificate.common.certificate.RecoveryDccV1
+import de.rki.coronawarnapp.covidcertificate.common.certificate.TestDccV1
+import de.rki.coronawarnapp.covidcertificate.common.certificate.VaccinationDccV1
+import de.rki.coronawarnapp.covidcertificate.common.exception.InvalidHealthCertificateException
+import de.rki.coronawarnapp.covidcertificate.common.exception.InvalidHealthCertificateException.ErrorCode.HC_DSC_EXPIRED
+import de.rki.coronawarnapp.covidcertificate.common.exception.InvalidHealthCertificateException.ErrorCode.HC_DSC_NOT_YET_VALID
+import de.rki.coronawarnapp.covidcertificate.common.exception.InvalidHealthCertificateException.ErrorCode.HC_DSC_NO_MATCH
+import de.rki.coronawarnapp.covidcertificate.common.exception.InvalidHealthCertificateException.ErrorCode.HC_DSC_OID_MISMATCH_RC
+import de.rki.coronawarnapp.covidcertificate.common.exception.InvalidHealthCertificateException.ErrorCode.HC_DSC_OID_MISMATCH_TC
+import de.rki.coronawarnapp.covidcertificate.common.exception.InvalidHealthCertificateException.ErrorCode.HC_DSC_OID_MISMATCH_VC
+import kotlinx.coroutines.flow.first
+import org.bouncycastle.asn1.ASN1Integer
+import org.bouncycastle.asn1.DERSequence
+import org.bouncycastle.asn1.pkcs.RSAPublicKey
+import org.bouncycastle.asn1.x509.SubjectPublicKeyInfo
+import org.bouncycastle.jce.provider.BouncyCastleProvider
 import timber.log.Timber
+import java.io.ByteArrayInputStream
+import java.math.BigInteger
+import java.security.KeyFactory
+import java.security.PublicKey
+import java.security.Security
+import java.security.Signature
+import java.security.cert.CertificateExpiredException
+import java.security.cert.CertificateFactory
+import java.security.cert.CertificateNotYetValidException
+import java.security.cert.X509Certificate
+import java.security.spec.RSAPublicKeySpec
 import javax.inject.Inject
 
 @Reusable
-class DscSignatureValidator @Inject constructor() {
+class DscSignatureValidator @Inject constructor(
+    private val dscRepository: DscRepository
+) {
 
-    suspend fun isSignatureValid(dscData: DscData, certificateData: DccData<*>): Boolean {
-        Timber.tag(TAG).d("isSignatureValid(certificateData=%s)", certificateData.header)
-        // TODO
-        return true
+    init {
+        Security.addProvider(BouncyCastleProvider()) // For SHA256withRSA/PSS
     }
 
+    private val vcOids = setOf(
+        "1.3.6.1.4.1.1847.2021.1.2",
+        "1.3.6.1.4.1.0.1847.2021.1.2"
+    )
+    private val tcOids = setOf(
+        "1.3.6.1.4.1.1847.2021.1.1",
+        "1.3.6.1.4.1.0.1847.2021.1.1"
+    )
+    private val rcOids = setOf(
+        "1.3.6.1.4.1.1847.2021.1.3",
+        "1.3.6.1.4.1.0.1847.2021.1.3"
+    )
+
+    private val oidSet = vcOids + tcOids + rcOids
+
+    /**
+     * @throws InvalidHealthCertificateException if validation fail, otherwise it is OK!
+     */
+    suspend fun validateSignature(dccData: DccData<*>) {
+        val dscData = dscRepository.dscData.first()
```

Comment:
@jurajkusnier  I have added `DscItem` which offers data a convenient types such as:

```kotlin
dscListItem -> DscItem(
  dscListItem.kid.toOkioByteString().base64(),
  dscListItem.data.toOkioByteString()
)
```

Please map `dscListItems` into it ?

Reply:
sure, it will be changed in my following PR